### PR TITLE
fix: blob.reload() does not work as intuitively expected

### LINKED ISSUE #149: Blob.reload() does not work as intuitively expected
Documentation for Blob.writer(BlobWriterOption...) recommends the following way to upload information:

```
 byte[] content = "Hello, World!".getBytes(UTF_8);
 try (WriteChannel writer = blob.writer()) {
   try {
     writer.write(ByteBuffer.wrap(content, 0, content.length));
   } catch (Exception ex) {
     // handle exception
   }
 }
```

But doc is silent on how to renew the blob object, which is a bit tricky because neither `blob.reload()` nor `storage.getBlob(blob.getBlobId())` does not help. Behavior varies depending on versioning setting for the bucket.

If versioning is enabled:

Attempt to reload as recommended:
`blob.reload(Blob.BlobSourceOption.generationNotMatch())` causes `StorageException '304 Not Modified'`

`blob.reload()`  returns the same blob.
`storage.get(blob.getBlobId())` returns the same blob (that seems very confusing).
And only `storage.get(blobId)` returns the updated blob.

If versioning is Suspended:
before upload: `blob.getContent()` returns an empty array
After upload: `blob.getContent()` throws `StorageException: 404 Not Found , No such object`
`storage.get(blob.getBlobId())`, `blob.reload()` and `blob.reload(Blob.BlobSourceOption.generationNotMatch())` return null.

And again only storage.get(blobId) returns the updated blob.

### PR DESCRIPTION
**Pull request by dmitry-fa**
Fixes #149

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java [hunk 1]
```diff
@@ -525,24 +525,38 @@ public boolean exists(BlobSourceOption... options) {
   }
 
   /**
-   * Fetches current blob's latest information. Returns {@code null} if the blob does not exist.
+   * Fetches the latest blob properties. Returns {@code null} if the blob no longer exists.
    *
-   * <p>Example of getting the blob's latest information, if its generation does not match the
-   * {@link Blob#getGeneration()} value, otherwise a {@link StorageException} is thrown.
+   * <p>{@code options} parameter can contain the preconditions. E.g. user may need to get the blob
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java:530]: E.g. --> For example,

Google style prohibits Latin abbreviations

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java [hunk 2]
```diff
@@ -525,24 +525,38 @@ public boolean exists(BlobSourceOption... options) {
...
-   * @param options blob read options
-   * @return a {@code Blob} object with latest information or {@code null} if not found
+   * @param options preconditions to fetch
+   * @return a {@code Blob} object with latest information or {@code null} if no longer exists.
    * @throws StorageException upon failure
    */
   public Blob reload(BlobSourceOption... options) {
-    return storage.get(getBlobId(), toGetOptions(this, options));
+    BlobId id = getBlobId();
+    return storage.get(BlobId.of(id.getBucket(), id.getName()), toGetOptions(this, options));
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java:559]: Why do you make a copy of the ID here?

Comment (dmitry-fa) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java:559]: BlobId includes three components: bucket, name and generation.  I form BlobId without generation to request the latest version of the blob.

Comment (elharo) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java:559]: OK, please add a comment explaining this

Comment (dmitry-fa) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: done

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java (hunk 1 above)
Comment (dmitry-fa) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: E.g. is widely used, perhaps we need to collect such things and do some code clean up.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java [hunk 3]
```diff
@@ -525,24 +525,39 @@ public boolean exists(BlobSourceOption... options) {
   }
 
   /**
-   * Fetches current blob's latest information. Returns {@code null} if the blob does not exist.
+   * Fetches the latest blob properties. Returns {@code null} if the blob no longer exists.
    *
-   * <p>Example of getting the blob's latest information, if its generation does not match the
-   * {@link Blob#getGeneration()} value, otherwise a {@link StorageException} is thrown.
+   * <p>{@code options} parameter can contain the preconditions. For example, user may need to get
```
Comment (elharo) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java:530]: user may need --> the user might want

Comment (dmitry-fa) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: done

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java [hunk 4]
```diff
@@ -525,24 +525,39 @@ public boolean exists(BlobSourceOption... options) {
...
+   *     // the content was updated externally
+   *   } else {
+   *     throw e;
+   *   }
    * }
    * }</pre>
    *
-   * @param options blob read options
-   * @return a {@code Blob} object with latest information or {@code null} if not found
+   * @param options preconditions to fetch
```
Comment (frankyn) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java:553]: preconditions to use, see https://cloud.google.com/storage/docs/json_api/v1/objects/get for more information.

Comment (dmitry-fa) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: reload-jdoc

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: Thanks @dmitry-fa,

use: `preconditions to use on reload, see https://cloud.google.com/storage/docs/json_api/v1/objects/get for more information` instead.

Comment (dmitry-fa) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: done

**Pull request merged**